### PR TITLE
Feature #710: Added game meta data file

### PR DESCRIPTION
**Pull request by sp0lsh**
All editor unit test passed

### CODE REVIEW

File: Assets/PatchKit Patcher/Scripts/AppData/Local/LocalMetaData.cs [hunk 1]
```diff
@@ -106,16 +142,17 @@ private void LoadData()
...
         {
-            _data = new Data
-            {
-                FileVersionIds = new Dictionary<string, int>()
-            };
+            _data = JsonConvert.DeserializeObject<Data>("{}"); // Json Deserializer will fill default property values defined in struct
+            _data.FileVersionIds = new Dictionary<string, int>();
+
+#if UNITY_5_3_OR_NEWER // LEGACY: fill productKey from unity prefs
+            _data.productKey = UnityEngine.PlayerPrefs.GetString(ValidateLicenseCommand.CachePatchkitKey, string.Empty);
```
Comment (genail) [inline Assets/PatchKit Patcher/Scripts/AppData/Local/LocalMetaData.cs]: 1. It's not a good place for loading license key from registry. LoadEmptyData() should fill data with empty values, it should't try and load any valid data even from previous location.
2. PlayerPrefs license key should be removed on successful load.

File: Assets/PatchKit Patcher/Scripts/AppData/Local/LocalMetaData.cs [hunk 2]
```diff
@@ -129,6 +166,12 @@ private bool TryLoadDataFromFile()
             try
             {
                 _data = JsonConvert.DeserializeObject<Data>(File.ReadAllText(_filePath));
+
+#if UNITY_5_3_OR_NEWER // LEGACY: fill productKey from unity prefs
+                _data.productKey = string.IsNullOrEmpty(_data.productKey)
+                    ? UnityEngine.PlayerPrefs.GetString(ValidateLicenseCommand.CachePatchkitKey, string.Empty)
```
Comment (genail) [inline Assets/PatchKit Patcher/Scripts/AppData/Local/LocalMetaData.cs]: This is actually a good place to load PlayerPrefs license key data. Just remember to remove it if present.

File: Assets/PatchKit Patcher/Scripts/AppData/Local/LocalMetaData.cs [hunk 3]
```diff
@@ -18,6 +20,22 @@ public class LocalMetaData : ILocalMetaData
         /// </summary>
         private struct Data
         {
+            [DefaultValue("patcher_data")]
+            [JsonProperty("file_id", DefaultValueHandling = DefaultValueHandling.Populate)]
+            public string fileId;
```
Comment (witcher112) [inline Assets/PatchKit Patcher/Scripts/AppData/Local/LocalMetaData.cs]: Should be `FileId`.

File: Assets/PatchKit Patcher/Scripts/AppData/Local/LocalMetaData.cs [hunk 4]
```diff
@@ -18,6 +20,22 @@ public class LocalMetaData : ILocalMetaData
...
+
+            [DefaultValue("1.0")]
+            [JsonProperty("version", DefaultValueHandling = DefaultValueHandling.Populate)]
+            public string version;
```
Comment (witcher112) [inline Assets/PatchKit Patcher/Scripts/AppData/Local/LocalMetaData.cs]: Should be `Version`.

File: Assets/PatchKit Patcher/Scripts/AppData/Local/LocalMetaData.cs [hunk 5]
```diff
@@ -18,6 +20,22 @@ public class LocalMetaData : ILocalMetaData
...
+
+            [DefaultValue("")]
+            [JsonProperty("product_key", DefaultValueHandling = DefaultValueHandling.Populate)]
+            public string productKey;
```
Comment (witcher112) [inline Assets/PatchKit Patcher/Scripts/AppData/Local/LocalMetaData.cs]: Should be `ProductKey`.

File: Assets/PatchKit Patcher/Scripts/AppData/Local/LocalMetaData.cs [hunk 6]
```diff
@@ -18,6 +20,22 @@ public class LocalMetaData : ILocalMetaData
...
+
+            [DefaultValue("none")]
+            [JsonProperty("product_key_encryption", DefaultValueHandling = DefaultValueHandling.Populate)]
+            public string productKeyEncryption;
+
```
Comment (witcher112) [inline Assets/PatchKit Patcher/Scripts/AppData/Local/LocalMetaData.cs]: Should be `ProductKeyEncryption`.

File: Assets/PatchKit Patcher/Scripts/AppUpdater/Commands/ValidateLicenseCommand.cs [hunk 7]
```diff
@@ -13,6 +13,8 @@ namespace PatchKit.Unity.Patcher.AppUpdater.Commands
     public class ValidateLicenseCommand : BaseAppUpdaterCommand, IValidateLicenseCommand
     {
         private static readonly DebugLogger DebugLogger = new DebugLogger(typeof(ValidateLicenseCommand));
+        public const string CachePatchkitKey = "patchkit-key";
```
Comment (witcher112) [inline Assets/PatchKit Patcher/Scripts/AppUpdater/Commands/ValidateLicenseCommand.cs]: Should be `CachePatchKitKey`.

File: Assets/PatchKit Patcher/Scripts/AppUpdater/Commands/ValidateLicenseCommand.cs [hunk 8]
```diff
@@ -13,6 +13,8 @@ namespace PatchKit.Unity.Patcher.AppUpdater.Commands
...
+        public const string CachePatchkitKeySecret = "patchkit-keysecret-";
```
Comment (witcher112) [inline Assets/PatchKit Patcher/Scripts/AppUpdater/Commands/ValidateLicenseCommand.cs]: Should be `CachePatchKitKeySecret`.

File: Assets/PatchKit Patcher/Scripts/App.cs [hunk 9]
```diff
@@ -163,7 +163,7 @@ private static IDownloadDirectory CreateDefaultDownloadDirectory(string appDataP
 
         private static ILocalMetaData CreateDefaultLocalMetaData(string appDataPath)
         {
-            return new LocalMetaData(appDataPath.PathCombine("patcher_cache.json"));
+            return new LocalMetaData(appDataPath.PathCombine("patcher_data.json"), appDataPath.PathCombine("patcher_cache.json"));
```
Comment (witcher112) [inline Assets/PatchKit Patcher/Scripts/App.cs]: Now I'm thinking whether shouldn't we call it `app_data.json`

What do you think @sp0lsh @genail ?

Comment (witcher112) [reply Assets/PatchKit Patcher/Scripts/App.cs]: @sp0lsh @genail bump

Comment (genail) [reply Assets/PatchKit Patcher/Scripts/App.cs]: Sorry @witcher112. Github notifications are one of the worst to notice...
Well, app_data.json sounds actually quite good. This file is for the app to access it, so only application-related data will be there.

Comment (witcher112) [reply Assets/PatchKit Patcher/Scripts/App.cs]: OK 👍 
@sp0lsh can you make such modification?

**Pull request merged**